src/sources/udp.rs

```diff
@@ -0,0 +1,220 @@
+use crate::{
+    event::{self, Event},
+    topology::config::{DataType, GlobalOptions, SourceConfig},
+};
+use codec::{self, BytesDelimitedCodec};
+use futures::{future, stream::iter_ok, sync::mpsc, Future, Sink, Stream};
+use serde::{Deserialize, Serialize};
+use std::{io, net::SocketAddr};
+use string_cache::DefaultAtom as Atom;
+use tokio::{
+    self,
+    codec::{BytesCodec, Decoder},
+    net::{UdpFramed, UdpSocket},
+};
+use tracing::field;
+
+/// UDP processes messages per packet, where messages are separated by newline.
+#[derive(Deserialize, Serialize, Debug, Clone)]
+#[serde(deny_unknown_fields)]
+pub struct UdpConfig {
+    pub address: SocketAddr,
+    pub host_key: Option<Atom>,
+}
+
+impl UdpConfig {
+    pub fn new(address: SocketAddr) -> Self {
+        Self {
+            address,
+            host_key: None,
+        }
+    }
+}
+
+#[typetag::serde(name = "udp")]
+impl SourceConfig for UdpConfig {
+    fn build(
+        &self,
+        _name: &str,
+        _globals: &GlobalOptions,
+        out: mpsc::Sender<Event>,
+    ) -> Result<super::Source, String> {
+        let host_key = self.host_key.clone().unwrap_or(event::HOST.clone());
+        Ok(udp(self.address, host_key, out))
+    }
+
+    fn output_type(&self) -> DataType {
+        DataType::Log
+    }
+}
+
+pub fn udp(address: SocketAddr, host_key: Atom, out: mpsc::Sender<Event>) -> super::Source {
+    let out = out.sink_map_err(|e| error!("error sending event: {:?}", e));
+
+    Box::new(
+        future::lazy(move || {
+            let socket = UdpSocket::bind(&address).expect("failed to bind to udp listener socket");
+
+            info!(message = "listening.", addr = &field::display(address),);
+
+            future::ok(socket)
```

Comment:
Take it or leave it 😄 , Result implements IntoFuture so you can just use it directly in closures.

```suggestion
            Ok(socket)
```